lib/student_greeter.rb

```diff
@@ -0,0 +1,84 @@
+class StudentGreeter
+  def self.greet_all_ungreeted_students
+    new.greet_all_ungreeted_students
+  end
+
+  def initialize
+    @greeters = User.where(greeter: true)
+  end
+
+  def greet_all_ungreeted_students
+    courses = Course.strictly_current
+    courses.each do |course|
+      nonstudents_in_course = course.nonstudents
+      possible_greeters = @greeters & nonstudents_in_course
+      next if possible_greeters.empty?
+      greeter = possible_greeters[0]
+      course.students_without_nonstudents.each do |student|
```

Comment:
Add the `ungreeted` flag here, so you don't need to do `greet_if_ungreeted` on the next line:

`course.students_without_nonstudents.ungreeted.collect { |s| greet(s) }`